### ISSUE TITLE
[Bug] middleware UndoManager

### ISSUE THREAD

**Issue by builden** [mobxjs/mobx-state-tree#457]
"mobx-state-tree": "^1.0.1"

```js
import { types as t, onSnapshot } from 'mobx-state-tree';
import UndoManager from 'mobx-state-tree/middleware/UndoManager';

const initObj = {
  obj: {
    arr: [{ a: 1, b: 2 }],
  },
};

const Mutable = t.union(t.string, t.boolean, t.number, t.map(t.late(() => Mutable)), t.array(t.late(() => Mutable)));
const Store = t
  .model('Store', {
    frozen: Mutable,
  })
  .actions(self => ({
    addProp(key, value) {
      self.frozen.set(key, value);
    },
  }));

const store = Store.create({ frozen: initObj });
const undoMgr = UndoManager.create({}, { targetStore: store });
store.addProp('a', { c: ['x'] });
```

undoMgr.history.length = 3; // expect 1
```json
// undoMgr.history
[
  {
    "patches": [],
    "inversePatches": []
  },
  {
    "patches": [],
    "inversePatches": []
  },
  {
    "patches": [
      {
        "op": "add",
        "path": "/frozen/a",
        "value": {
          "c": [
            "x"
          ]
        }
      }
    ],
    "inversePatches": [
      {
        "op": "remove",
        "path": "/frozen/a"
      }
    ]
  }
]
```

if I test, `store.addProp(b, 1)`, can get the right result, undoMgr.history.length = 1

**Comment by mweststrate** [comment]
@builden thanks for reporting! Would you mind creating a PR with this as unit test? I suspect it is an issue in handling nested actions.

**Comment by builden** [comment]
of cause you can do it

**Comment by builden** [comment]
of cause you can do it :)

**Comment by mweststrate** [comment]
Should be fixed in 1.3. Please reopen if it isn't the case
